src/rowbinary/de.rs

```diff
@@ -308,43 +396,245 @@ impl<'data> Deserializer<'data> for &mut RowBinaryDeserializer<'_, 'data> {
         panic!("tuple struct types are unsupported: `{name}`");
     }
 
-    #[inline]
+    #[inline(always)]
     fn deserialize_ignored_any<V: Visitor<'data>>(self, _visitor: V) -> Result<V::Value> {
         panic!("ignored types are unsupported");
     }
 
-    #[inline]
+    #[inline(always)]
     fn is_human_readable(&self) -> bool {
         false
     }
 }
 
-fn get_unsigned_leb128(mut buffer: impl Buf) -> Result<u64> {
-    let mut value = 0u64;
-    let mut shift = 0;
+/// Used in [`Deserializer::deserialize_seq`], [`Deserializer::deserialize_tuple`],
+/// and it could be used in [`Deserializer::deserialize_struct`],
+/// if we detect that the field order matches the database schema.
+struct RowBinarySeqAccess<'de, 'cursor, 'data, R: Row, Validator>
+where
+    Validator: SchemaValidator<R>,
+{
+    deserializer: &'de mut RowBinaryDeserializer<'cursor, 'data, R, Validator>,
+    len: usize,
+}
+
+impl<'data, R: Row, Validator> SeqAccess<'data> for RowBinarySeqAccess<'_, '_, 'data, R, Validator>
+where
+    Validator: SchemaValidator<R>,
+{
+    type Error = Error;
+
+    fn next_element_seed<T>(&mut self, seed: T) -> Result<Option<T::Value>>
+    where
+        T: DeserializeSeed<'data>,
+    {
+        if self.len > 0 {
+            self.len -= 1;
+            let value = DeserializeSeed::deserialize(seed, &mut *self.deserializer)?;
+            Ok(Some(value))
+        } else {
+            Ok(None)
+        }
+    }
+
+    fn size_hint(&self) -> Option<usize> {
+        Some(self.len)
+    }
+}
 
-    loop {
-        ensure_size(&mut buffer, 1)?;
+/// Used in [`Deserializer::deserialize_map`].
+struct RowBinaryMapAccess<'de, 'cursor, 'data, R: Row, Validator>
+where
+    Validator: SchemaValidator<R>,
+{
+    deserializer: &'de mut RowBinaryDeserializer<'cursor, 'data, R, Validator>,
+    entries_visited: usize,
+    len: usize,
```

Comment:
It seems that one counter is enough ("Returns the number of entries remaining in the map" 
 in https://docs.rs/serde/latest/serde/de/trait.MapAccess.html#method.size_hint)